index.js

```diff
@@ -0,0 +1,693 @@
+              var types = $typeVarMap[typeVarName];
+              okTypes = chain(types, function(t) {
+                return all(values, t.test) ? [t] : [];
+              });
+              if (isEmpty(okTypes)) {
+                throw invalidArgument(name, map(types, nest), _value, index);
+              }
+            } else {
+              okTypes = chain(commonTypes(map(values, determineActualTypes)),
+                              rejectInconsistent);
+              if (isEmpty(okTypes) && !isEmpty(values)) {
+                throw orphanArgument(env, name, _value, index);
+              }
+            }
+            if (!isEmpty(okTypes)) $typeVarMap[typeVarName] = okTypes;
+            return okTypes;
+
+          case 'UNARY':
+            $1s = recur(expType.$1,
+                        chain(values, expType._1),
+                        UnaryType.from(expType));
+            return map(or($1s, [expType.$1]), UnaryType.from(expType));
+
+          case 'BINARY':
+            var specialize = BinaryType.from(expType);
+            $1s = recur(expType.$1,
+                        chain(values, expType._1),
+                        function($1) { return specialize($1, expType.$2); });
+            $2s = recur(expType.$2,
+                        chain(values, expType._2),
+                        function($2) { return specialize(expType.$1, $2); });
+            return BinaryType.xprod(expType,
+                                    or($1s, [expType.$1]),
+                                    or($2s, [expType.$2]));
+
+          default:
+            return commonTypes(map(values, determineActualTypes));
+        }
+      };
+    };
+
+    var satisfactoryTypes =
+    function(name, constraints, $typeVarMap, expType, value, index) {
+      return _satisfactoryTypes(name, constraints, $typeVarMap, value, index)
+                               (expType, [value], id);
+    };
+
+    var curry = function(name, constraints, expArgTypes, expRetType,
+                         _typeVarMap, _values, _indexes, impl) {
+      return arity(_indexes.length, function() {
+        var delta = _indexes.length - arguments.length;
+        if (delta < 0) {
+          throw new TypeError(
+            LEFT_SINGLE_QUOTATION_MARK + name + RIGHT_SINGLE_QUOTATION_MARK +
+            ' requires ' + numArgs(expArgTypes.length) + ';' +
+            ' received ' + numArgs(expArgTypes.length - delta)
+          );
+        }
+        var $typeVarMap = {};
+        for (var typeVarName in _typeVarMap) {
+          $typeVarMap[typeVarName] = _typeVarMap[typeVarName].slice();
+        }
+        var values = _values.slice();
+        var indexes = [];
+        for (var i = 0; i < _indexes.length; i += 1) {
+          var index = _indexes[i];
+
+          if (i < arguments.length &&
+              !(typeof arguments[i] === 'object' &&
+                arguments[i] != null &&
+                arguments[i]['@@functional/placeholder'] === true)) {
```

Comment:
Interesting idea! I'm not sure it's a good idea, though. A placeholder represents the absence of a value, so treating a placeholder as a value could be a source of confusion. Also, we'd still need to check for a placeholder here explicitly, so we wouldn't actually leverage the existing type-checking machinery.